dragonfly/actions/action_text.py

```diff
@@ -3,109 +3,178 @@
-#   under the terms of the GNU Lesser General Public License as published 
-#   by the Free Software Foundation, either version 3 of the License, or 
+#   Dragonfly is free software: you can redistribute it and/or modify it
+#   under the terms of the GNU Lesser General Public License as published
+#   by the Free Software Foundation, either version 3 of the License, or
 #   (at your option) any later version.
 #
-#   Dragonfly is distributed in the hope that it will be useful, but 
-#   WITHOUT ANY WARRANTY; without even the implied warranty of 
-#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU 
+#   Dragonfly is distributed in the hope that it will be useful, but
+#   WITHOUT ANY WARRANTY; without even the implied warranty of
+#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 #   Lesser General Public License for more details.
 #
-#   You should have received a copy of the GNU Lesser General Public 
-#   License along with Dragonfly.  If not, see 
+#   You should have received a copy of the GNU Lesser General Public
+#   License along with Dragonfly.  If not, see
 #   <http://www.gnu.org/licenses/>.
 #
-
 """
 Text action
 ============================================================================
 
-This section describes the :class:`Text` action object. This type of 
+This section describes the :class:`Text` action object. This type of
 action is used for typing text into the foreground application.
 
-It differs from the :class:`Key` action in that :class:`Text` is used for 
-typing literal text, while :class:`dragonfly.actions.action_key.Key` 
-emulates pressing keys on the keyboard.  An example of this is that the 
-arrow-keys are not part of a text and so cannot be typed using the 
-:class:`Text` action, but can be sent by the 
+It differs from the :class:`Key` action in that :class:`Text` is used for
+typing literal text, while :class:`dragonfly.actions.action_key.Key`
+emulates pressing keys on the keyboard.  An example of this is that the
+arrow-keys are not part of a text and so cannot be typed using the
+:class:`Text` action, but can be sent by the
 :class:`dragonfly.actions.action_key.Key` action.
 
 """
 
 
-from .action_base           import DynStrActionBase, ActionError
-from .typeables             import typeables
-from .keyboard              import Keyboard
-from .action_key            import Key
-from ..windows.clipboard    import Clipboard
-from ..engines              import get_engine
+from six import text_type
+
+from ..engines import get_engine
+from ..windows.clipboard import Clipboard
+from ..windows.window import Window
+from .action_base import ActionError, DynStrActionBase
+from .action_key import Key
+from .keyboard import Keyboard
+from .typeables import typeables
+
+# ---------------------------------------------------------------------------
+
+UNICODE_KEYBOARD = True
+HARDWARE_APPS = [
+            "tvnviewer.exe", "vncviewer.exe", "mstsc.exe", "virtualbox.exe"
```

Comment:
I'm operating under the assumption these programs simply echo the virtual keys they receive and the system Windows uses for Unicode emulation is simply nonsensical outside the context of the host system. See below for details.

Reply:
Makes sense. I read up a bit more about how all this works. For anyone following along, here's how I understand it: the basic question I wanted to understand is how someone who uses a language that requires non-ASCII characters would use VirtualBox to type into a guest OS. The answer is that they would have a non-US keyboard layout on both their host system and the guest system. The virtual keys that are typed would be the same in both host and guest, and these are independent of keyboard layout. Then, the host and guest system would separately translate these using os-dependent mechanisms into non-ASCII characters based on their non-US keyboard layouts.

This new Dragonfly functionality allows anyone with any arbitrary keyboard layout to type any Unicode character. This functionality cannot be expressed using this low-level virtual key system. Instead, it uses a Windows-specific system for Unicode emulation, which isn't going to work on the guest. There's not really any way around this, because there isn't an OS-independent standard for specifying Unicode keypresses.